src/v/datalake/partition_key_path.cc

```diff
@@ -0,0 +1,324 @@
+
+    std::chrono::system_clock::time_point tp{std::chrono::milliseconds(0)};
+    // offset epoch by the given duration
+    tp += DurationT(v.val);
+
+    if constexpr (std::is_same_v<std::chrono::years, DurationT>) {
+        return ssx::sformat("{:%Y}", tp);
+    }
+
+    if constexpr (std::is_same_v<std::chrono::months, DurationT>) {
+        return ssx::sformat("{:%Y-%m}", tp);
+    }
+
+    if constexpr (std::is_same_v<std::chrono::days, DurationT>) {
+        return ssx::sformat("{:%Y-%m-%d}", tp);
+    }
+
+    if constexpr (std::is_same_v<std::chrono::hours, DurationT>) {
+        return ssx::sformat("{:%Y-%m-%d-%H}", tp);
+    }
+}
+
+struct transform_value_formatting_visitor {
+    explicit transform_value_formatting_visitor(
+      const iceberg::primitive_value& v)
+      : value(v) {}
+    // This case handles:
+    // - Identity transform
+    // - Truncation transform
+    checked<ss::sstring, partition_key_error> operator()(const auto&) {
+        return std::visit(primitive_formatting_visitor{}, value);
+    }
+
+    checked<ss::sstring, partition_key_error>
+    operator()(const iceberg::year_transform&) {
+        return format_time_transform_key<std::chrono::years>(value);
+    }
+
+    checked<ss::sstring, partition_key_error>
+    operator()(const iceberg::month_transform&) {
+        return format_time_transform_key<std::chrono::months>(value);
+    }
+    checked<ss::sstring, partition_key_error>
+    operator()(const iceberg::day_transform&) {
+        return format_time_transform_key<std::chrono::days>(value);
+    }
+    checked<ss::sstring, partition_key_error>
+    operator()(const iceberg::hour_transform&) {
+        return format_time_transform_key<std::chrono::hours>(value);
+    }
+    checked<ss::sstring, partition_key_error>
+    operator()(const iceberg::void_transform&) {
+        return "null";
+    }
+
+    const iceberg::primitive_value& value;
+};
+
+struct value_formatting_visitor {
+    checked<ss::sstring, partition_key_error>
+    operator()(const iceberg::primitive_value& v) {
+        return std::visit(primitive_formatting_visitor{}, v);
+    }
+    checked<ss::sstring, partition_key_error> operator()(const auto&) {
+        return partition_key_error(
+          "non primitive iceberg partition values are not supported");
+    }
+};
+
+ss::sstring escape(std::string_view s) {
+    static constexpr std::
+      array<std::pair<absl::string_view, absl::string_view>, 25>
+        encode_lut = {
+          {{"&", "%26"},   {"$", "%24"},  {"@", "%40"}, {"=", "%3D"},
+           {";", "%3B"},   {"/", "%2F"},  {":", "%3A"}, {"+", "%2B"},
+           {" ", "%20"},   {",", "%2C"},  {"?", "%3F"}, {"\"", "%22"},
+           {"#", "%23"},   {"%", "%25"},  {"<", "%3C"}, {">", "%3E"},
+           {"[", "%5B"},   {"\\", "%5C"}, {"]", "%5D"}, {"^", "%5E"},
+           {"`", "%60"},   {"{", "%7B"},  {"|", "%7C"}, {"}", "%7D"},
+           {"\177", "%7F"}}};
+
+    return absl::StrReplaceAll(s, encode_lut);
```

Comment:
Questions: 
- absl::StrReplaceAll is O(n*m)?
- Should this be a simple loop over `s` instead with byte-copying and for specific bytes appending the percent encoded variant instead?
- The function assumes valid utf-8 input? Should we have an extra check for validating input as it doesn't seem to handle broken encodings, right? Or we leave that to the caller? Let's document the preconditions if so.
- - Related: Where does "\177" come from?

- tried looking at ada and their encode function is marked as private
- there is boost::url which has an encode method but we don't include it...
- we use boost::beast but from 5 minute glancing i couldn't find if/where it handles encoding

Reply:
i was looking for url encoding function but didn't find one that we could, do you think the complexity is a concern here ?

Reply:
it is in the `internal` namespace, also have some limitations